File: bin/backupagent.py
import os
import sys
import socket
import zmq
import subprocess
import msgpack
import psutil
import signal
import logging
from time import sleep


class BackupAgent:
    def __init__(self):
        self.agentid = self.get_agent_id()
        self.master = "offsite.tacticaltechs.com"
        self.zmq_conn = f"tcp://{self.master}:23956"
        self.veeam_exe = os.path.join(
            "C:\\Program Files\\Veeam\\Endpoint Backup", "Veeam.EndPoint.Manager.exe"
        )
        logging.basicConfig(
            filename=os.path.join(os.getcwd(), "backup.log"),
            level=logging.INFO,
            format="%(asctime)s - %(name)s - %(levelname)s - %(message)s",
        )
        self.logger = logging.getLogger(__name__)

    def get_agent_id(self):
        cmd = ["wmic", "csproduct", "get", "uuid"]
        r = subprocess.run(cmd, capture_output=True)
        wmic = r.stdout.decode().splitlines()[2].strip()
        hostname = socket.gethostname()
        return f"{hostname}|{wmic}"

    def kill_proc(self, pid):
        try:
            parent = psutil.Process(pid)
            children = parent.children(recursive=True)
            children.append(parent)
            for p in children:
                p.send_signal(signal.SIGTERM)

            gone, alive = psutil.wait_procs(children, timeout=20, callback=None)
        except Exception:
            pass

    def start_backup(self, mode):
        cmd = [self.veeam_exe, mode]
        r = subprocess.Popen(cmd, stdin=None, stdout=None, stderr=None, close_fds=True)
        self.logger.info(f"Backup started with pid {r.pid}")
        return r.pid

    def backups_running(self):

        pids = []

        for proc in psutil.process_iter():
            with proc.oneshot():
                if proc.name() == "Veeam.EndPoint.Manager.exe":
                    pids.append(proc.pid)

        return pids

    def get_procs(self):
        ret = []

        for proc in psutil.process_iter():
            i = {}
            try:
                with proc.oneshot():
                    if proc.name() == "Veeam.EndPoint.Manager.exe":
                        i["name"] = proc.name()
                        i["pid"] = proc.pid
                        i["ppid"] = proc.ppid()
                        i["cmdline"] = proc.cmdline()
                        ret.append(i)
            except Exception:
                continue

        return ret

    def run(self):
        self.logger.info("Backup agent starting")

        context = zmq.Context()

        sock = context.socket(zmq.DEALER)
        sock.setsockopt(zmq.LINGER, 0)
        sock.setsockopt(zmq.TCP_KEEPALIVE, 1)
        sock.setsockopt(zmq.TCP_KEEPALIVE_IDLE, 300)
        sock.setsockopt(zmq.TCP_KEEPALIVE_INTVL, 300)
        sock.setsockopt_string(zmq.IDENTITY, self.agentid)
        sock.connect(self.zmq_conn)

        poll = zmq.Poller()
        poll.register(sock, zmq.POLLIN)

        while 1:
            try:
                sockets = dict(poll.poll(1000))
                
                if sock in sockets:
                    msg = sock.recv()
                    data = msgpack.loads(msg)

                    if data and data["target"] == self.agentid:

                        if data["cmd"] == "startbackup":
                            # make sure backup is not already running
                            running = self.backups_running()

                            if running:
                                payload = {"ret": "failed"}
                            else:
                                pid = self.start_backup(data["mode"])

                                proc = psutil.Process(pid)

                                with proc.oneshot():
                                    name = proc.name()
                                    cmdline = proc.cmdline()

                                payload = {
                                    "ret": "success",
                                    "pid": pid,
                                    "proc_name": name,
                                    "cmdline": cmdline,
                                }

                            payload["id"] = self.agentid
                            sock.send(msgpack.dumps(payload))

                        elif data["cmd"] == "cancelbackup":

                            try:
                                proc = psutil.Process(data["pid"])
                            except psutil.NoSuchProcess:
                                payload = {"ret": "failed"}
                            else:
                                self.kill_proc(data["pid"])
                                self.logger.info(
                                    f"Cancelled backup with pid: {data['pid']}"
                                )
                                payload = {"ret": "success"}

                            payload["id"] = self.agentid
                            sock.send(msgpack.dumps(payload))

                        elif data["cmd"] == "info":
                            procs = self.get_procs()
                            payload = {"ret": "success", "procs": procs, "id": self.agentid}
                            sock.send(msgpack.dumps(payload))

                        elif data["cmd"] == "ping":
                            payload = {
                                "ret": "success",
                                "msg": f"{self.agentid} pong",
                                "id": self.agentid,
                            }
                            sock.send(msgpack.dumps(payload))

                        elif data["cmd"] == "halt":
                            payload = {"ret": "success", "id": self.agentid}
                            sock.send(msgpack.dumps(payload))
                            self.logger.info("Shutting down")
                            break

                        del msg
            except Exception as e:
                self.logger.error(e)
                break
        
        sock.close()
        context.term()


def main():
    agent = BackupAgent()
    agent.run()


if __name__ == "__main__":
    main()

File: bin/backupagent.go
//go:generate goversioninfo -icon=onit.ico -manifest=goversioninfo.exe.manifest -gofile=versioninfo.go
package main

import (
	"flag"
	"fmt"
	nats "github.com/nats-io/nats.go"
	"github.com/shirou/gopsutil/process"
	"github.com/ugorji/go/codec"
	"log"
	"os"
	"os/exec"
	"path/filepath"
	"runtime"
	"strings"
	"time"
)

var (
	debuglog *os.File
	logger   *log.Logger
	endpoint string
	token    string
	version  = "1.2.0"
)

type proc struct {
	Name    string   `json:"name"`
	Pid     int32    `json:"pid"`
	Ppid    int32    `json:"ppid"`
	Cmdline []string `json:"cmdline"`
}

type procList []proc

type simpleRet struct {
	AgentID string `json:"id"`
	Ret     string `json:"ret"`
}

type backupRet struct {
	AgentID string   `json:"id"`
	Ret     string   `json:"ret"`
	Pid     int      `json:"pid"`
	Name    string   `json:"proc_name"`
	Cmdline []string `json:"cmdline"`
}

type infoRet struct {
	AgentID string   `json:"id"`
	Ret     string   `json:"ret"`
	Procs   procList `json:"procs"`
}

type pongRet struct {
	AgentID string `json:"id"`
	Ret     string `json:"ret"`
	Payload string `json:"msg"`
}

func main() {
	ver := flag.Bool("version", false, "Prints version")
	flag.Parse()

	if *ver {
		fmt.Println(fmt.Sprintf("Tactical Backup %s", version))
		return
	}

	setupLogging()
	defer debuglog.Close()

	logger.Println("Backup agent starting...")

	agentid := getAgentID()

	opts := []nats.Option{nats.Name("Tactical Backup Agent"), nats.Token(token)}
	opts = setupConnOptions(opts)

	server := fmt.Sprintf("nats://%s:4222", endpoint)
	nc, err := nats.Connect(server, opts...)
	if err != nil {
		logger.Println(err)
	}

	var data map[string]string
	var response []byte

	nc.Subscribe(agentid, func(msg *nats.Msg) {

		setupLogging()
		dec := codec.NewDecoderBytes(msg.Data, new(codec.MsgpackHandle))
		if err := dec.Decode(&data); err != nil {
			logger.Fatal(err)
		}

		if data["cmd"] == "startbackup" {

			if backupIsRunning() {

				ret := codec.NewEncoderBytes(&response, new(codec.MsgpackHandle))
				ret.Encode(simpleRet{AgentID: agentid, Ret: "failed"})
				msg.Respond(response)

			} else {

				pid := startBackup(data["mode"])
				logger.Println(fmt.Sprintf("Backup started with pid %d", pid))

				p, _ := process.NewProcess(int32(pid))
				name, _ := p.Name()
				cmdline, _ := p.CmdlineSlice()

				ret := codec.NewEncoderBytes(&response, new(codec.MsgpackHandle))
				ret.Encode(backupRet{AgentID: agentid, Ret: "success", Pid: pid, Name: name, Cmdline: cmdline})
				msg.Respond(response)

			}
		} else if data["cmd"] == "info" {

			procs := getProcs()

			ret := codec.NewEncoderBytes(&response, new(codec.MsgpackHandle))
			ret.Encode(infoRet{Ret: "success", AgentID: agentid, Procs: procs})
			msg.Respond(response)

		} else if data["cmd"] == "ping" {

			payload := fmt.Sprintf("%s pong", agentid)
			ret := codec.NewEncoderBytes(&response, new(codec.MsgpackHandle))
			ret.Encode(pongRet{Ret: "success", AgentID: agentid, Payload: payload})
			msg.Respond(response)

		} else if data["cmd"] == "halt" {

			logger.Println("Shutting down")

			ret := codec.NewEncoderBytes(&response, new(codec.MsgpackHandle))
			ret.Encode(simpleRet{Ret: "success", AgentID: agentid})
			msg.Respond(response)

			nc.Flush()
			nc.Close()
			os.Exit(0)
		}

	})
	nc.Flush()

	if err := nc.LastError(); err != nil {
		log.Fatal(err)
	}

	runtime.Goexit()
}

func getAgentID() string {
	out, _ := exec.Command("wmic", "csproduct", "get", "uuid").Output()
	hostname, _ := os.Hostname()
	uid := strings.Split(string(out), "\n")[1]
	uid = strings.TrimSpace(uid)

	return fmt.Sprintf("%s|%s", hostname, uid)
}

func startBackup(mode string) int {
	veeam := filepath.Join("C:\\Program Files\\Veeam\\Endpoint Backup", "Veeam.EndPoint.Manager.exe")
	cmd := exec.Command(veeam)
	cmd.Start()

	return cmd.Process.Pid
}

func backupIsRunning() bool {
	pids, _ := process.Pids()

	for _, pid := range pids {
		p, _ := process.NewProcess(pid)
		name, _ := p.Name()

		if name == "Veeam.EndPoint.Manager.exe" {
			return true
		}
	}

	return false
}

func getProcs() procList {
	ret := make([]proc, 0)
	pids, _ := process.Pids()

	for _, pid := range pids {

		p, _ := process.NewProcess(pid)
		name, _ := p.Name()
		pid := p.Pid
		ppid, _ := p.Ppid()
		cmdline, _ := p.CmdlineSlice()

		if name == "Veeam.EndPoint.Manager.exe" {
			var i = proc{
				name,
				pid,
				ppid,
				cmdline,
			}
			ret = append(ret, i)
		}
	}
	return ret
}

func setupLogging() {
	debuglog, _ = os.OpenFile("backupagent.log", os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
	logger = log.New(debuglog, "", log.LstdFlags)
}

func setupConnOptions(opts []nats.Option) []nats.Option {
	opts = append(opts, nats.ReconnectWait(time.Second*5))
	opts = append(opts, nats.RetryOnFailedConnect(true))
	opts = append(opts, nats.MaxReconnects(-1))
	return opts
}
